### ISSUE TITLE
libtpm2-pk11 segfaults

### ISSUE THREAD

**Issue by svenschwermer** [irtimmer/tpm2-pk11#57]
I have compiled and installed the latest master branches from tpm2-tss, tpm2-abrmd and tpm2-tools. They talk to an IBM soft-TPM. I had to slightly modify tpm2-pk11 to make it compatible with the new TCTI device and socket API (https://github.com/irtimmer/tpm2-pk11/compare/master...svenschwermer:master). The tpm2-tools work as expected.
When I run `p11tool --provider=libtpm2-pk11.so --list-privkeys`, however, libtpm2-pk11 segfaults in `attr_get`. Running the command with a debug build from gdb with logging enabled yields:
```
─── Output/messages
2018-03-07 21:38:44 [tpm-pk11] C_GetInfo
2018-03-07 21:38:44 [tpm-pk11] C_GetSlotList: present = true
2018-03-07 21:38:44 [tpm-pk11] C_GetTokenInfo: id = 4660
2018-03-07 21:38:44 [tpm-pk11] C_GetSlotInfo: id = 4660
2018-03-07 21:38:44 [tpm-pk11] C_OpenSession: id = 4660, flags = 4
2018-03-07 21:38:44 [tpm-pk11] C_CloseSession: session = 55800e30
2018-03-07 21:38:44 [tpm-pk11] C_GetSlotList: present = true
2018-03-07 21:38:44 [tpm-pk11] C_GetTokenInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_GetSlotInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_OpenSession: id = 4660, flags = 4
2018-03-07 21:38:45 [tpm-pk11] C_CloseSession: session = e400b8c0
2018-03-07 21:38:45 [tpm-pk11] C_GetSlotList: present = true
2018-03-07 21:38:45 [tpm-pk11] C_GetTokenInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_GetSlotInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_OpenSession: id = 4660, flags = 4
2018-03-07 21:38:45 [tpm-pk11] C_CloseSession: session = e400bb70
2018-03-07 21:38:45 [tpm-pk11] C_GetSlotList: present = true
2018-03-07 21:38:45 [tpm-pk11] C_GetTokenInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_GetSlotInfo: id = 4660
2018-03-07 21:38:45 [tpm-pk11] C_OpenSession: id = 4660, flags = 4
2018-03-07 21:38:45 [tpm-pk11] C_FindObjectsInit: session = e400be00, count = 1
2018-03-07 21:38:45 [tpm-pk11] C_FindObjects: session = e400be00, max = 8192

Thread 1 "p11tool" received signal SIGSEGV, Segmentation fault.
─── Assembly
0x00007ffff5ca4fb3 attr_get+277 mov    -0x18(%rbp),%eax
0x00007ffff5ca4fb6 attr_get+280 movslq %eax,%rdx
0x00007ffff5ca4fb9 attr_get+283 mov    -0x28(%rbp),%rax
0x00007ffff5ca4fbd attr_get+287 mov    0x18(%rax),%rax
0x00007ffff5ca4fc1 attr_get+291 cmp    %rax,%rdx
0x00007ffff5ca4fc4 attr_get+294 jb     0x7ffff5ca4eba <attr_get+28>
0x00007ffff5ca4fca attr_get+300 mov    $0x0,%eax
─── Source
48   attr_dynamic_index_of(CKA_SERIAL_NUMBER, PkcsX509, serial, serial_size),
49   attr_index_of(CKA_CERTIFICATE_TYPE, PkcsX509, cert_type),
50 };
51 
52 void* attr_get(pObject object, CK_ATTRIBUTE_TYPE type, size_t *size) {
53   for (int i = 0; i < object->num_entries; i++) {
54     pAttrIndexEntry entries = &object->entries[i];
55     for (int j = 0; j < entries->num_attrs; j++) {
56       if (type == entries->indexes[j].type) {
57         pAttrIndex index = &entries->indexes[j];
58         if (index->size_offset == 0) {
─── Stack
[0] from 0x00007ffff5ca4fbd in attr_get+287 at /home/sven/tpm2/tpm2-pk11/src/object.c:53
arg object = 0x0
arg type = 0
arg size = 0x7fffffffcf60
[1] from 0x00007ffff5ca63f2 in C_FindObjects+277 at /home/sven/tpm2/tpm2-pk11/src/pk11.c:181
arg session_handle = 140737018641920
arg object_handle = 0x5555558094b0
arg max_objects = 8192
arg found = 0x7fffffffd408
[+]
───
0x00007ffff5ca4fbd in attr_get (object=0x0, type=0, size=0x7fffffffcf60) at /home/sven/tpm2/tpm2-pk11/src/object.c:53
53	  for (int i = 0; i < object->num_entries; i++) {
```
We can see that `attr_get` is called with `object=0`.

**Comment by svenschwermer** [comment]
I just tried the same with the following versions:

- tpm2-tss 1.3.0
- tpm2-abrmd 1.2.0
- tpm2-tools 3.0.3
- tpm2-pk11 at 0876bca

The same thing happens.

**Comment by svenschwermer** [comment]
I managed to make everything work on a different machine (Ubuntu 16.04 vs. Arch which I used before). However, I do run into a segfault in the working setup when the TPM doesn't hold any keys (remanufactured) and then run `p11tool --list-keys`. Perhaps this was the reason for the crash on the other machine as well.

**Comment by irtimmer** [comment]
It indeed crashes when there are no keys. I already have a patch to fix it, but haven't it committed and pushed it.

**Comment by irtimmer** [comment]
Fixed with commit 2f0ed78f33d09b401f2392b34c12dae0152e5519
